diff --git a/node_modules/scratch-vm/src/compiler/compat-blocks.js b/node_modules/scratch-vm/src/compiler/compat-blocks.js
index 86f03f26..e438bb20 100644
--- a/node_modules/scratch-vm/src/compiler/compat-blocks.js
+++ b/node_modules/scratch-vm/src/compiler/compat-blocks.js
@@ -10,10 +10,8 @@
 const stacked = [
     'looks_changestretchby',
     'looks_hideallsprites',
-    'looks_sayforsecs',
     'looks_setstretchto',
     'looks_switchbackdroptoandwait',
-    'looks_thinkforsecs',
     'motion_align_scene',
     'motion_glidesecstoxy',
     'motion_glideto',
@@ -21,7 +19,6 @@ const stacked = [
     'motion_pointtowards',
     'motion_scroll_right',
     'motion_scroll_up',
-    'sensing_askandwait',
     'sensing_setdragmode',
     'sound_changeeffectby',
     'sound_changevolumeby',
diff --git a/node_modules/scratch-vm/src/compiler/enums.js b/node_modules/scratch-vm/src/compiler/enums.js
index 1d493bf4..2f798dda 100644
--- a/node_modules/scratch-vm/src/compiler/enums.js
+++ b/node_modules/scratch-vm/src/compiler/enums.js
@@ -180,6 +180,7 @@ const StackOpcode = {
     PEN_STAMP: 'pen.stamp',
 
     SENSING_TIMER_RESET: 'timer.reset',
+    SENSING_ASK: 'sensing.ask',
 
     PROCEDURE_RETURN: 'procedures.return',
     PROCEDURE_CALL: 'procedures.call'
diff --git a/node_modules/scratch-vm/src/compiler/irgen.js b/node_modules/scratch-vm/src/compiler/irgen.js
index 08d1701f..a029a442 100644
--- a/node_modules/scratch-vm/src/compiler/irgen.js
+++ b/node_modules/scratch-vm/src/compiler/irgen.js
@@ -691,6 +691,10 @@ class ScriptTreeGenerator {
             return new IntermediateStackBlock(StackOpcode.CONTROL_WAIT_UNTIL, {
                 condition: this.descendInputOfBlock(block, 'CONDITION').toType(InputType.BOOLEAN)
             }, true);
+        case 'sensing_askandwait':
+            return new IntermediateStackBlock(StackOpcode.SENSING_ASK, {
+                question: this.descendInputOfBlock(block, 'QUESTION')
+            });
         case 'control_while':
             return new IntermediateStackBlock(StackOpcode.CONTROL_WHILE, {
                 condition: this.descendInputOfBlock(block, 'CONDITION').toType(InputType.BOOLEAN),
@@ -809,6 +813,7 @@ class ScriptTreeGenerator {
         case 'looks_nextcostume':
             return new IntermediateStackBlock(StackOpcode.LOOKS_COSTUME_NEXT);
         case 'looks_say':
+        case 'looks_sayforsecs':
             return new IntermediateStackBlock(StackOpcode.LOOKS_SAY, {
                 message: this.descendInputOfBlock(block, 'MESSAGE')
             });
@@ -832,6 +837,7 @@ class ScriptTreeGenerator {
                 costume: this.descendInputOfBlock(block, 'COSTUME', true)
             });
         case 'looks_think':
+        case 'looks_thinkforsecs':
             return new IntermediateStackBlock(StackOpcode.LOOKS_THINK, {
                 message: this.descendInputOfBlock(block, 'MESSAGE')
             });
@@ -1113,7 +1119,7 @@ class ScriptTreeGenerator {
                 variant,
                 arguments: args
             },
-            yields: !this.script.isWarp && procedureCode === this.script.procedureCode
+            yields: false
         };
     }
 
@@ -1318,7 +1324,7 @@ class ScriptTreeGenerator {
     }
 
     analyzeLoop () {
-        return !this.script.isWarp || this.script.warpTimer;
+        return false;
     }
 
     readTopBlockComment (commentId) {
diff --git a/node_modules/scratch-vm/src/compiler/jsexecute.js b/node_modules/scratch-vm/src/compiler/jsexecute.js
index 61d69cd3..642f0809 100644
--- a/node_modules/scratch-vm/src/compiler/jsexecute.js
+++ b/node_modules/scratch-vm/src/compiler/jsexecute.js
@@ -459,7 +459,6 @@ runtimeFunctions.listReplace = `const listReplace = (list, idx, value) => {
         return;
     }
     list.value[index] = value;
-    list._monitorUpToDate = false;
 }`;
 
 /**
@@ -474,7 +473,6 @@ runtimeFunctions.listInsert = `const listInsert = (list, idx, value) => {
         return;
     }
     list.value.splice(index, 0, value);
-    list._monitorUpToDate = false;
 }`;
 
 /**
@@ -492,7 +490,6 @@ runtimeFunctions.listDelete = `const listDelete = (list, idx) => {
         return;
     }
     list.value.splice(index, 1);
-    list._monitorUpToDate = false;
 }`;
 
 /**
diff --git a/node_modules/scratch-vm/src/compiler/jsgen.js b/node_modules/scratch-vm/src/compiler/jsgen.js
index ebbdc6bf..dab098c5 100644
--- a/node_modules/scratch-vm/src/compiler/jsgen.js
+++ b/node_modules/scratch-vm/src/compiler/jsgen.js
@@ -534,7 +534,6 @@ class JSGenerator {
                 }
                 this.source += '}\n'; // close switch
                 this.source += `if (!${branchVariable}.isLoop) break;\n`;
-                this.yieldLoop();
                 this.source += '}\n'; // close while
             } else {
                 throw new Error(`Unknown block type: ${blockType}`);
@@ -593,7 +592,6 @@ class JSGenerator {
             this.source += `${index}++; `;
             this.source += `${this.referenceVariable(node.variable)}.value = ${index};\n`;
             this.descendStack(node.do, new Frame(true));
-            this.yieldLoop();
             this.source += '}\n';
             break;
         }
@@ -616,7 +614,6 @@ class JSGenerator {
                 this.source += `for (var ${i} = ${this.descendInput(node.times)}; ${i} >= 0.5; ${i}--) {\n`;
             }
             this.descendStack(node.do, new Frame(true));
-            this.yieldLoop();
             this.source += `}\n`;
             break;
         }
@@ -652,11 +649,6 @@ class JSGenerator {
         case StackOpcode.CONTROL_WHILE:
             this.source += `while (${this.descendInput(node.condition)}) {\n`;
             this.descendStack(node.do, new Frame(true));
-            if (node.warpTimer) {
-                this.yieldStuckOrNotWarp();
-            } else {
-                this.yieldLoop();
-            }
             this.source += `}\n`;
             break;
         case StackOpcode.CONTROL_CLEAR_COUNTER:
@@ -677,19 +669,16 @@ class JSGenerator {
         case StackOpcode.LIST_ADD: {
             const list = this.referenceVariable(node.list);
             this.source += `${list}.value.push(${this.descendInput(node.item)});\n`;
-            this.source += `${list}._monitorUpToDate = false;\n`;
             break;
         }
         case StackOpcode.LIST_DELETE: {
             const list = this.referenceVariable(node.list);
             if (node.index.isConstant('last')) {
                 this.source += `${list}.value.pop();\n`;
-                this.source += `${list}._monitorUpToDate = false;\n`;
                 break;
             }
             if (node.index.isConstant(1)) {
                 this.source += `${list}.value.shift();\n`;
-                this.source += `${list}._monitorUpToDate = false;\n`;
                 break;
             }
             // do not need a special case for all as that is handled in IR generation (list.deleteAll)
@@ -707,7 +696,6 @@ class JSGenerator {
             const item = this.descendInput(node.item);
             if (node.index.isConstant(1)) {
                 this.source += `${list}.value.unshift(${item});\n`;
-                this.source += `${list}._monitorUpToDate = false;\n`;
                 break;
             }
             this.source += `listInsert(${list}, ${this.descendInput(node.index)}, ${item});\n`;
@@ -780,10 +768,13 @@ class JSGenerator {
             this.source += `runtime.ext_scratch3_looks._setCostume(target, ${this.descendInput(node.costume)});\n`;
             break;
         case StackOpcode.LOOKS_SAY:
-            this.source += `runtime.ext_scratch3_looks._say(${this.descendInput(node.message)}, target);\n`;
+            this.source += `runtime._scratch_run_say(${this.descendInput(node.message.toType(InputType.STRING))});\n`;
             break;
         case StackOpcode.LOOKS_THINK:
-            this.source += `runtime.ext_scratch3_looks._think(${this.descendInput(node.message)}, target);\n`;
+            this.source += `runtime._scratch_run_think(${this.descendInput(node.message.toType(InputType.STRING))});\n`;
+            break;
+        case StackOpcode.SENSING_ASK:
+            this.source += `runtime.ext_scratch3_sensing._answer = runtime._scratch_run_ask(${this.descendInput(node.question.toType(InputType.STRING))});\n`;
             break;
 
         case StackOpcode.MOTION_X_CHANGE:
@@ -861,18 +852,12 @@ class JSGenerator {
             break;
 
         case StackOpcode.PROCEDURE_CALL: {
-            const procedureCode = node.code;
             const procedureVariant = node.variant;
             const procedureData = this.ir.procedures[procedureVariant];
             if (procedureData.stack === null) {
                 // TODO still need to evaluate arguments
                 break;
             }
-            const yieldForRecursion = !this.isWarp && procedureCode === this.script.procedureCode;
-            if (yieldForRecursion) {
-                // Direct yields.
-                this.yieldNotWarp();
-            }
             if (procedureData.yields) {
                 this.source += 'yield* ';
                 if (!this.script.yields) {
diff --git a/node_modules/scratch-vm/src/engine/runtime.js b/node_modules/scratch-vm/src/engine/runtime.js
index 1df8dff1..a2c48d7d 100644
--- a/node_modules/scratch-vm/src/engine/runtime.js
+++ b/node_modules/scratch-vm/src/engine/runtime.js
@@ -2523,6 +2523,7 @@ class Runtime extends EventEmitter {
         // Clean up threads that were told to stop during or since the last step
         this.threads = this.threads.filter(thread => !thread.isKilled);
         this.updateThreadMap();
+        this._emitProjectRunStatus(this.threads.length - this._getMonitorThreadCount(this.threads));
 
         // Find all edge-activated hats, and add them to threads to be evaluated.
         for (const hatType in this._hats) {
@@ -2533,7 +2534,6 @@ class Runtime extends EventEmitter {
             }
         }
         this.redrawRequested = false;
-        this._pushMonitors();
         if (this.profiler !== null) {
             if (stepThreadsProfilerId === -1) {
                 stepThreadsProfilerId = this.profiler.idByName('Sequencer.stepThreads');
